content/blog/2025-12-16-bracket.md

```diff
@@ -0,0 +1,261 @@
++++
+title = "[bracket]: Bridging Bril and Racket Educational Compilers"
+[extra]
+bio = """
+  Joseph is an ECE PhD student at Cornell :)
+"""
+[[extra.authors]]
+name = "Joseph Maheshe"
++++
+
+**Code:** [https://github.com/maheshejs/cs6120pr](https://github.com/maheshejs/cs6120pr)
+
+## What was the goal?
+
+The goal of this project was to connect two educationally-focused intermediate languages: Racket's exprs-lang-v9 subset from UBC's CPSC411 (Compiler Construction) course and Bril, the Big Red Intermediate Language. While Bril already has TypeScript and Rust frontends, these support only limited subsets of their source languages. Racket's exprs-lang-v9, in contrast, is a more expressive subset that includes first-class procedures, making it an ideal candidate for generating interesting Bril benchmarks. The project aimed to compile exprs-lang-v9 programs—which normally compile all the way down to x64 machine code into Bril programs using the dynamic type extension.
+
+## What did I do?
+
+### Design Overview
+
+My compiler takes programs in `proc-imp-cmf-lang-v7` (an intermediate representation in the CPSC411 compiler pipeline) and lowers them to Bril JSON. This particular IR was chosen because it sits at the sweet spot after all high-level abstractions have been lowered but before architecture-specific concerns like register allocation.
+
+Choosing this intermediate language was itself a significant design decision. Initially, I considered starting very close to assembly since Bril is close to assembly. However, x64 and lower-level languages in the CPSC411 pipeline allow copying a label to a location (treating code as data), which Bril does not support. This would have required designing an abstract interpreter to uncover labels in locations and abstract physical locations to abstract locations, plus extensive instruction patching. That turned out to be cumbersome.
+
+Instead, I chose `proc-imp-cmf-lang-v7`, which perfectly satisfied my requirements:
+- It already uses abstract locations (before register allocation and spillage)
+- It hasn't yet transformed calls into low-level calling conventions
+- Its predicates can be easily transformed to basic blocks with appropriate labels
+- Calls can be directly translated to Bril's function calls
```

Comment:
This is a very clear summary for someone not really familiar with the IR; thanks!